keps/sig-network/2104-reworking-kube-proxy-architecture/README.md

```diff
@@ -0,0 +1,1111 @@
+      to implement technical details of writing backend routing rules
+      from services to pods (eg. iptables, ipvs, Windows kernel).
+
+    - A gRPC API for optimal communication between the core logic
+      daemon and the backend implementations, which can be run in memory
+      or externalized (so that the brain can run with a one-many relationship
+      with "proxy backends"
+
+    - A set of golang packages for working with the gRPC API, which third parties
+      can use to create their own proxy backend implementations out of tree.
+
+- Provide an implementation of the core service proxy logic daemon.
+  (This implementation will be used, unmodified, by all core and
+  third-party proxy backend implementations.)
+
+- Provide a golang library that consumes the gRPC API from the above
+  daemon and exports the information to the proxy backend
+  implementation in one of two forms:
+
+    - a "full state" model, in which the library keeps track of the
+      full state of Kubernetes network statespace on this node, and
+      sends the proxy implementation a new copy of the full state on every
+      update. The library will also include a package called
+      "diffstore" that is designed to make it very easy for
+      implementations to generate incremental updates based on the
+      full state information.
+
+    - an "incremental" model, in which the library simply passes on
+      the updates from the gRPC API so the proxy implementation can
+      update its own model, which allows for similar behaviour to the
+      current upstream kube-proxy.
+
+- Provide additional reusable library elements for some proxy logic,
+  such as conntrack cleaning, which might be called from different
+  proxy implementations.
+
+- Provide new implementations of the existing "standard" proxy
+  implementations (iptables, ipvs, and Windows kernel), based on the
+  new daemon and client library. At least one of these will use the
+  "full state" model and at least one will use the "incremental" model.
+
+- Deprecate and eventually remove the existing proxy implementations
+  in `k8s.io/kubernetes`, in favor of the new implementations. Also
+  remove the associated support packages in `k8s.io/kubernetes` that
+  are only used by kube-proxy (eg, `pkg/util/ipvs`,
+  `pkg/util/conntrack`, `pkg/util/netsh`).
+
+### Non-Goals
+
+- We Won't necessarily provide bulletproof NFT, eBPF, Userspace backends with parity to the core Windows kernel, IPTAbles, IPVS implementations.  
+- We Won't Require KPNG to run in a mode where the kpng "core logic brain" is a separate process from the kpng "proxy backends", we state this as non goal because it's been a bit of a red herring debate in the past.  Although KPNG support this, it's not required. 
+- We won't Require all proxiers to use the fullstate model.  However, this is ideal for new implementations, we think because it's easier to read and understand.
```

Comment:
do you see a future where the same proxy can work for both k8s, istio and/or linkerd ? Is that a goal which would require cross collaboration ?

Reply:
no :) ... we are attempting to constrain the goals of this kep for now, but its worth talking in another thread